dev/bots/analyze_framework_code.dart

```diff
@@ -0,0 +1,421 @@
+      ):
+        // Expressions such as `final int x = 1.clamp(0, 2);` should be allowed.
+        // Do nothing.
+        break;
+      case MethodInvocation(
+        target: Expression(staticType: DartType(isDartCoreDouble: true) || DartType(isDartCoreNum: true) || DartType(isDartCoreInt: true)),
+      ):
+        clampAccessNodes.add(node);
+      case _:
+        break;
+    }
+  }
+}
+
+// ----------- Verify No _debugAssert -----------
+
+final ResolvedUnitVerifier verifyDebugAssertAccess = _DebugAssertVerifier();
+class _DebugAssertVerifier extends ResolvedUnitVerifier {
+  final List<String> accessErrors = <String>[];
+  final List<String> annotationErrors = <String>[];
+  final Map<ExecutableElement, bool> lookup = <ExecutableElement, bool>{};
+  @override
+  void reportError() {
+    if (annotationErrors.isNotEmpty) {
+      foundError(<String>[
+        '${bold}Overriding a framework class member that was not annotated with @_debugAssert and marking the override @_debugAssert is not allowed.$reset',
+        '${bold}A framework method/getter/setter not marked as debug-only itself cannot have a debug-only override.$reset\n',
+        ...annotationErrors,
+        '\n${bold}Consider either removing the @_debugAssert annotation, or adding the annotation to the class member that is being overridden instead.$reset',
+      ]);
+    }
+    if (accessErrors.isNotEmpty) {
+      foundError(<String>[
+        '${bold}Framework symbols annotated with @_debugAssert should not be accessed outside of asserts.$reset\n',
+        ...accessErrors,
+      ]);
+    }
+  }
+
+  @override
+  void analyzeResolvedUnitResult(ResolvedUnitResult unit, String workingDirectory) {
+    if (unit.libraryElement.metadata.any(_DebugAssertVisitor._isDebugAssertAnnotationElement)) {
+      return;
+    }
+    final _DebugAssertVisitor visitor = _DebugAssertVisitor(lookup);
+    unit.unit.visitChildren(visitor);
+    final String relativePath = path.relative(unit.path, from: workingDirectory);
+    for (final (AstNode node, Element element) in visitor.accessViolations) {
+      final LineInfo lineInfo = unit.lineInfo;
+      final int lineNumber = lineInfo.getLocation(node.offset).lineNumber;
+      final String name = switch (element) {
+        ConstructorElement(:final InterfaceElement enclosingElement, name: '') => '${enclosingElement.name}.new',
+        ExecutableElement(:final InterfaceElement enclosingElement, :final String name) => '${enclosingElement.name}.$name',
+        Element(:final String? name) => name ?? '',
+      };
+      accessErrors.add('$relativePath:$lineNumber: $bold$name$reset accessed outside of an assert.');
+    }
+
+    for (final (String superImpl, String override) in visitor.incorrectAnnotations) {
+      annotationErrors.add('$relativePath: class member $bold$superImpl$reset is not annotated wtih $bold@_debugAssert$reset, but its override $bold$override$reset is.');
+    }
```

Comment:
It would be easy to create a new rule to enforce this. But what if someone wants to mark an unnamed constructor as debug-only (or an operator)?

Reply:
unnamed constructor -> then the class should probably be called DebugFoo, so the rule would be fulfilled again. 

debug-only operator: We should just disallow this. The naming rule is that anything only to be called in debug mode should begin with debug, and well, you cannot define an operator that does that. It would be somewhat confusing to allow this.

See also https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#qualify-variables-and-methods-used-only-for-debugging